### ISSUE TITLE
Model using polymorphic relation throw exception

### ISSUE THREAD

**Issue by warlof** [yajra/laravel-datatables#2573]
### Summary of problem or feature request

Hi, have some model using polymorphic relations that I'm calling with Datatable factory. It appears that sorting and filtering data on them are causing exceptions.

Are they not supported at time ?

This is the exception I get while doing sorting on a column :

```
[2021-02-07 21:55:22] local.ERROR: SQLSTATE[42000]: Syntax error or access violation: 1066 Not unique table/alias: 'character_wallet_transactions' (SQL: select * from `character_wallet_transactions` left join `character_wallet_transactions` on `character_wallet_transactions`.`location_id` = `character_wallet_transactions`.`` where `character_wallet_transactions`.`character_id` in (94768465) order by `character_wallet_transactions`.`name` asc limit 10 offset 0) {"userId":1,"exception":"[object] (Yajra\\DataTables\\Exceptions\\Exception(code: 0): SQLSTATE[42000]: Syntax error or access violation: 1066 Not unique table/alias: 'character_wallet_transactions' (SQL: select * from `character_wallet_transactions` left join `character_wallet_transactions` on `character_wallet_transactions`.`location_id` = `character_wallet_transactions`.`` where `character_wallet_transactions`.`character_id` in (94768465) order by `character_wallet_transactions`.`name` asc limit 10 offset 0) at /var/www/seat/vendor/yajra/laravel-datatables-oracle/src/DataTableAbstract.php:862)
[stacktrace]
#0 /var/www/seat/vendor/yajra/laravel-datatables-oracle/src/QueryDataTable.php(106): Yajra\\DataTables\\DataTableAbstract->errorResponse(Object(Illuminate\\Database\\QueryException))
#1 /var/www/seat/packages/eveseat/web/src/Http/DataTables/Common/Financial/AbstractWalletTransactionDataTable.php(94): Yajra\\DataTables\\QueryDataTable->make(true)
#2 /var/www/seat/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Seat\\Web\\Http\\DataTables\\Common\\Financial\\AbstractWalletTransactionDataTable->ajax()
#3 /var/www/seat/vendor/laravel/framework/src/Illuminate/Container/Util.php(37): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#4 /var/www/seat/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure(Object(Closure))
#5 /var/www/seat/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\\Container\\BoundMethod::callBoundMethod(Object(Illuminate\\Foundation\\Application), Array, Object(Closure))
#6 /var/www/seat/vendor/laravel/framework/src/Illuminate/Container/Container.php(590): Illuminate\\Container\\BoundMethod::call(Object(Illuminate\\Foundation\\Application), Array, Array, NULL)
#7 /var/www/seat/vendor/yajra/laravel-datatables-buttons/src/Services/DataTable.php(158): Illuminate\\Container\\Container->call(Array)
```

This is the exception I get while doing some search :

```
[2021-02-07 21:55:26] local.ERROR: Please use whereHasMorph() for MorphTo relationships. {"userId":1,"exception":"[object] (Yajra\\DataTables\\Exceptions\\Exception(code: 0): Please use whereHasMorph() for MorphTo relationships. at /var/www/seat/vendor/yajra/laravel-datatables-oracle/src/DataTableAbstract.php:862)
[stacktrace]
#0 /var/www/seat/vendor/yajra/laravel-datatables-oracle/src/QueryDataTable.php(106): Yajra\\DataTables\\DataTableAbstract->errorResponse(Object(RuntimeException))
#1 /var/www/seat/packages/eveseat/web/src/Http/DataTables/Common/Financial/AbstractWalletTransactionDataTable.php(94): Yajra\\DataTables\\QueryDataTable->make(true)
#2 /var/www/seat/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Seat\\Web\\Http\\DataTables\\Common\\Financial\\AbstractWalletTransactionDataTable->ajax()
```

### Code snippet of problem

Here is the main model used by the factory : `CharacterWalletTransaction`

```php
class CharacterWalletTransaction extends Model
{
    /**
     * @var array
     */
    protected $casts = [
        'is_buy' => 'boolean',
        'is_personal' => 'boolean',
    ];

    /**
     * @var array
     */
    protected $hidden = ['character_id', 'client_id', 'type_id', 'created_at', 'updated_at'];

    /**
     * @var bool
     */
    protected static $unguarded = true;

    /**
     * @var string
     */
    protected $primaryKey = 'id';

    /**
     * @var bool
     */
    public $incrementing = true;

    /**
     * @return \Illuminate\Database\Eloquent\Relations\MorphTo
     */
    public function location()
    {
        return $this->morphTo()->withDefault();
    }
}
```

One of the possible morphed model : `UniverseStructure`

```php
class UniverseStructure extends UniverseLocation
{

    /**
     * @var bool
     */
    protected static $unguarded = true;

    /**
     * @var bool
     */
    public $incrementing = false;

    /**
     * @var string
     */
    protected $primaryKey = 'structure_id';
}
```

The datatable factory (relaying on an abstract class since there are two way to render similar table)

https://github.com/eveseat/web/blob/master/src/Http/DataTables/Character/Financial/WalletTransactionDataTable.php
https://github.com/eveseat/web/blob/master/src/Http/DataTables/Common/Financial/AbstractWalletTransactionDataTable.php

### System details

- Operating System : Windows using an alpine container
- PHP Version : 7.3.20
- Laravel Version : 6.20.12
- Laravel-Datatables : 9.15.2

**Comment by parsasamandi** [comment]
->filterColumn('status', function ($query, $keyword) { 
                &nbsp; &nbsp; $statuses = Course::whereHas('statuses',  function ($subquery) use ($keyword) {
                    &nbsp; &nbsp; &nbsp; &nbsp; $subquery->where('status', $keyword);
               &nbsp; &nbsp; &nbsp; })->get()->pluck('id')->toArray();
                &nbsp; &nbsp; &nbsp; $query->whereIn('id', $statuses);
  })

Imaging there is a 1 to 1 relationship between course and status.
`
    public function statuses() {
        return $this->morphOne('App\Models\Status', 'status');
 }`
`
 public function status() {
        return $this->morphTo();
 }
`
.
use Column::make('status') in `getColumns()` function.
hope it works.

**Comment by warlof** [comment]
Hi,

Tried but it doesn't work as expected.
I also have an issue with sorting (which make sense afaict as I assume we have to provide a kind of union query to make it usable by the factory.)

I've been able to work on polymorphic relationship implementation for filter scope without too much troubles.

If anyone got an idea to also make them sortable, that would be great :)

**Comment by parsasamandi** [comment]
> Hi,
> 
> Tried but it doesn't work as expected.
> I also have an issue with sorting (which make sense afaict as I assume we have to provide a kind of union query to make it usable by the factory.)
> 
> I've been able to work on polymorphic relationship implementation for filter scope without too much troubles.
> 
> If anyone got an idea to also make them sortable, that would be great :)

I do not underestand what you mean by sorting. the default one?
In addition, the way you have implemented you dataTable is a little different than me, so maybe the reason behind not working is that.

**Comment by yajra** [comment]
Fixed via #2580. Thanks!
